flash/core/trainer.py

```diff
@@ -277,14 +279,31 @@ def request_dataloader(
             The dataloader
         """
         model, stage, is_legacy = self._parse_request_dataloader_args(args, kwargs)
+
         if is_legacy:
             self.call_hook(f"on_{stage}_dataloader")
             dataloader = getattr(model, f"{stage}_dataloader")()
         else:
             hook = f"{stage.dataloader_prefix}_dataloader"
             self.call_hook("on_" + hook, pl_module=model)
-            dataloader = self.call_hook(hook, pl_module=model)
+
+            dataloader = None
+            if _PL_GREATER_EQUAL_1_5_0:
+                source = getattr(self._data_connector, f"_{stage.dataloader_prefix}_dataloader_source")
+                if (
+                    not source.is_module()
+                    or not isinstance(source.instance, DataModule)
+                    or isinstance(source.instance, (LightningModule, NewDataModule))
+                ):
+                    dataloader = source.dataloader()
+
+            if dataloader is None:
```

Comment:
why only do this if the dataloader is None?

Reply:
This would request the dataloader if it hasn't been loaded yet. This would happen in custom loops I believe.

Reply:
Refactored it to make it clearer :smiley: